### PR TITLE
Add CloudFormation resource type configuration support

### PR DESCRIPTION
**Pull request by rajdivya**
*Issue #, if available:*

*Description of changes:*
Add TypeConfiguration support for Resource Types

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ammokhov — COMMENTED** [review]
this must be merged befor #369

**Review by ammokhov — CHANGES_REQUESTED** [review]
plz fix coverage

**Comment by rajdivya** [comment]
Failing checks are due to unreleased changes in cfn-cli. Validated that they pass locally.

```
➜  cloudformation-cli-java-plugin git:(typeConfig) pre-commit run --all-files

[WARNING] The 'rev' field of repo 'https://github.com/ambv/black' appears to be a mutable reference (moving tag / branch).  Mutable references are never updated after first install and are not supported.  See https://pre-commit.com/#using-the-latest-version-for-a-repository for more details.  Hint: `pre-commit autoupdate` often fixes this.
isort....................................................................Passed
black....................................................................^[[1;9CPassed
Check for case conflicts.................................................Passed
Fix End of Files.........................................................Passed
Mixed line ending........................................................Passed
Trim Trailing Whitespace.................................................Passed
Flake8...................................................................Passed
Pretty format JSON.......................................................Passed
Check for merge conflicts................................................Passed
Check Yaml...............................................................Passed
Check blanket noqa.......................................................Passed
check for not-real mock methods..........................................Passed
use logger.warning(......................................................Passed
bandit...................................................................Passed
pylint-local.............................................................Passed
pytest-local.............................................................Passed
```

**Comment by rajdivya** [comment]
Validated type configuration changes manually by running `cfn-init`, `cfn-generate` and `cfn-validate`.

File: src/test/java/software/amazon/cloudformation/exceptions/CfnInvalidTypeConfigurationExceptionTests.java [hunk 1]
```diff
@@ -9,7 +9,7 @@
     public void cfnInvalidTypeConfigurationException_isBaseHandlerException() {
         assertThatExceptionOfType(BaseHandlerException.class).isThrownBy(() -> {
             throw new CfnInvalidTypeConfigurationException("AWS::Type::Resource", "<request>");
-        }).withCauseInstanceOf(RuntimeException.class).withMessageContaining("<request>")
+        }).withNoCause().withMessageContaining("<request>")
```
Comment (omkhegde) [inline src/test/java/software/amazon/cloudformation/exceptions/CfnInvalidTypeConfigurationExceptionTests.java:12]: @ammokhov is this change correct?

**Comment by omkhegde** [comment]
Let's also remove the version bump changes from this PR to have a clean history

**Review by omkhegde — CHANGES_REQUESTED** [review]
Let's remove the plugin version bump from this PR to maintain clean history

File: src/test/java/software/amazon/cloudformation/exceptions/CfnInvalidTypeConfigurationExceptionTests.java (hunk 1 above)
Comment (ammokhov) [reply src/test/java/software/amazon/cloudformation/exceptions/CfnInvalidTypeConfigurationExceptionTests.java:12]: looks ok to me

**Review by omkhegde — APPROVED** [review]

**Review by anshikg — APPROVED** [review]

File: src/main/java/software/amazon/cloudformation/scheduler/CloudWatchScheduler.java [hunk 2]
```diff
@@ -81,9 +81,10 @@ public void refreshClient() {
      * @param <ResourceT> resource model state configuration to apply
      * @param <CallbackT> callback context associated with reschedule context
      */
-    public <ResourceT, CallbackT> void rescheduleAfterMinutes(final String functionArn,
-                                                              final int minutesFromNow,
-                                                              final HandlerRequest<ResourceT, CallbackT> handlerRequest) {
+    public <ResourceT, CallbackT,
+        ConfigurationT> void rescheduleAfterMinutes(final String functionArn,
+                                                    final int minutesFromNow,
+                                                    final HandlerRequest<ResourceT, CallbackT, ConfigurationT> handlerRequest) {
```
Comment (ammokhov) [inline src/main/java/software/amazon/cloudformation/scheduler/CloudWatchScheduler.java:87]: is it even used anywhere? @srujithpoondla03

**Review by ammokhov — APPROVED** [review]

File: src/main/java/software/amazon/cloudformation/scheduler/CloudWatchScheduler.java (hunk 2 above)
Comment (srujithpoondla03) [reply src/main/java/software/amazon/cloudformation/scheduler/CloudWatchScheduler.java:87]: Nope, I think there is some cleanup required in this repo for V1 Code(not just this pull request). We did not totally remove the code related to CW events yet.

**Pull request merged**